### ISSUE TITLE
Add a Dijkstra flag to cPath

### ISSUE THREAD

**Issue by SafwatHalaby** [cuberite/cuberite#2775]
This is just a potential improvement.
A* is good for going to a specific, known coordinate.
Dijkstra is good for finding a nearest block that satisfy some codition. e.g. Finding the nearest wheat block with some meta for a villager. Or finding the nearest dark spot for a zombie. Or finding the nearest water block for a burning mob.

It is trivial to make cPath work like that by adding a flag which switches from A* to Dijkstra and proper `cPath::FindBlockBy(...)` methods.

**Comment by madmaxoft** [comment]
I'm not sure if actually having multiple "pathfinding" classes with a common interface work better. That way, we can have A*, Dijkstra, A* with wallclimbing (for spiders), water-only A* (for guardians) etc.

**Comment by SafwatHalaby** [comment]
@madmaxoft The thing is, the difference between  `A*`, `Dijkstra`, `A* with wallclimbing` and `water-only A*` is extremely small. Wouldn't it be cleaner to simply add a flag to the cPath constructor and create a new dijkstra constructor?

```
enum eType { WALLCLIMBER, WATERONLY };
cPath::cPath(...., eType type);
cPath::cPath(<Block type and meta>);
```

It even logically makes sense: Either create a cPath and pass it the coords, or create a cPath and pass it the block you're after.

**Comment by SphinxC0re** [comment]
And what if you aren't searching for a single block but for a block structure

**Comment by SafwatHalaby** [comment]
Good point. The methods would need to be slightly different. Maybe have a destination-evaluating functor for Dijkstra. That'd also eradicate having a special case for light level.

```
cPath::cPath(<A functor>);
MyFunctor(cChunk & chunk, coords)
{
   Evaluate stuff as you wish and decide if this is your destination.
}
```

**Comment by pwnorbitals** [comment]
pathFinding code has seen updates over the years I presume. Is this issue still relevant ?

**Comment by tigerw** [comment]
Don't think it's been implemented, and I'd say mob AI still needs a lot of work before we'd need this.
